torch/_subclasses/functional_tensor.py

```diff
@@ -299,6 +299,9 @@ def __init__(self, pre_dispatch=False, export=False, _allow_token_discovery=Fals
         # track of the ordering between side effectful operations.
         self._tokens: Dict[Any, torch.Tensor] = {}
 
+        # Filled after forward tracing.
+        self._tokens_forward_output: Dict[Any, torch.Tensor] = {}
```

Comment:
Hmm I was going to say it would be nice if we don't need to add any state at all to `FunctionalTensorMode`. But I guess if you remove this, you'll need another way to plumb the list of forward-only tokens up the `joint_fn()` to the `handle_effect_tokens_fn`, so maybe sticking a field here isn't so bad.

Reply:
oh, looking at your code changes in `handle_effect_tokens_fn`, you technically don't need to do this: you can just grab the entire list of tokens after tracing the joint, and assume that if `len(functional_mode._tokens) > len(fw_metadata.tokens)`, the extra tokens on our functional mode correspond to backward tokens.

If you want to keep the assert it that sounds ok, but if so I would add a comment here describing the situation (that the only reason we're storing this extra state on `FunctionalTensorMode` is to more easily add asserts about the number of fw vs bw tokens).

Reply:
If we keep them inside _tokens for backward, they will be reused in backward.
So we need some place to store them and to add to the outs of runtime.
Do you mean to add just [None] * num_tokens instead of real result values?

Reply:
resolved offline - we're using `FunctionalTensorMode` here as a global object to stash the forward tokens, so that later in the token-wrapper-fn we know which tokens are fw_outs and bw_outs (a comment describing the situation in `FunctionTensorMode` would be great)